### ISSUE TITLE
2.1 AuthZ improvements

### ISSUE THREAD

**Issue by HaoK** [aspnet/Security#1359]
This issue will track all the various Authorization improvements we are looking at for 2.1.

## Some initial thoughts:

From @davidfowl 

We've had a bunch of feedback about our authz system with respect to flowing context from the authorize attribute to the authorization handler:
 
- https://github.com/aspnet/Security/issues/670
- https://github.com/aspnet/Mvc/issues/5607
- https://github.com/aspnet/Security/issues/917
- https://github.com/aspnet/Security/issues/1352

Today the Authorize attribute supports has enough metadata to describe the policy but it doesn't let you specify the resource (via IAuthorizeData). I think if we solve that, it might solve a bulk of the issues since people will be able to write custom attributes that flow the relevant context to the authorization handler. 
 
Today that's only possible when doing imperative authz. I was thinking something like:
 
```
public interface IAuthorizeResource
{
    object Resource { get; }
}
```
 
If the attribute implemented this, we would flow that as the resource to the handler. This way you implement permissions of whatever you please via resources.
 
From @rynowak 

Look into providing overriding semantics in MVC, maybe a marker interface: `IAuthorizeMetadata`. Any attributes that implement that interface on the endpoint could be flowed. Then it’s up to developers to build whatever they want.  (Also look into flowing single objects vs many objects + requirements VS resources)

We will also look into making it possible to specify requirements via Attributes similar to imperative AuthZ so you don't have to preconstruct policies for attributes.

**Comment by rynowak** [comment]
My additional thoughts on this. We're looking at making the concept of an **endpoint** available at a low level from the dispatcher. Endpoints will have arbitrary metadata in an ordered list. This is the equivalent to what MVC does today by passing the action descriptor.

-------------------------------------

Additionally in the case of MVC, the current `[Authorize]` has some surprising behaviors compared to other filters. All authorize filters will execute rather than supporting overriding.  Additionally, `[AllowAnonymous]` cannot be overridden and turns off **all** authorize attributes. These might be consistent with the past, but they don't really fit with the *zen* of MVC filters.

**Comment by HaoK** [comment]
Interesting, well if we have a new endpoint concept with a clean slate for metadata, we should probably just expose all of the basic building blocks 

1. Building the ClaimsPrincipal, need to end up with 0 or more authentication schemes.  If none are specified, we continue to use the default user aka httpContext.User, otherwise we build a merged user via authenticating all of the schemes.  We could manifest this with a series of [Authenticate(scheme)] attributes.  

2. Specifying the Authorization requirements, today the metadata only targets policies, but we could have [Authorize(typeof(MyRequirement), typeof(MyOtherRequirement)]

Some examples:

`[Authorize]` 

Would behave the same as today (using some default policy which relies on context.User, and still only does RequireAuthenticatedUser by default

`[Authenticate("Bearer")]`
`[Authorize]`

Would use the Bearer scheme and only require that it exists.

`[Authenticate("Bearer", "Cookie")]`
`[Authorize(typeof(RequireAdminClaim)]`

Would use the merged Bearer Cookie schemes and require an admin claim.

**Comment by Tratcher** [comment]
Which challenge would you issue?

**Comment by HaoK** [comment]
All of them, its not any different than today

**Comment by Tratcher** [comment]
Challenging both Bearer and Cookie is nonsensical.

**Comment by HaoK** [comment]
Not disagreeing, but today if you have a policy that has Bearer and Cookies, that's what it does today as well

**Comment by HaoK** [comment]
Unless we want to eliminate being able to specify multiple schemes entirely...  that's an option that would simplify things...

**Comment by davidfowl** [comment]
> Not disagreeing, but today if you have a policy that has Bearer and Cookies, that's what it does today as well

It does? What happens as a result? Don't they stomp on each other?

**Comment by HaoK** [comment]
Yeah they do, last one wins, multiple auth schemes in policies has always been a bit weird

**Comment by Tratcher** [comment]
Some are compatible like Bearer, Basic, and Windows.

**Comment by Eilon** [comment]
@HaoK please log new issues for any 2.1.0 proposals.

**Comment by hmvs** [comment]
@HaoK Is it going to be a part of 2.1 release?

**Comment by hmvs** [comment]
I want to have ability to return custom json if specific policy failed. for example 
403
{
"errorType": "PrivalcyPolicyAcceptRequired",
"errorMessage": "You token doesn't have PolicyAccepted claim"
}

**Comment by HaoK** [comment]
No the error authZ improvements didn't make it into 2.1, I'll try to get them in 2.2

**Comment by spottedmahn** [comment]
I assume that is why the title changed to 2.2 😜⁉

**Comment by HaoK** [comment]
So rough idea of some initial AuthZ improvements I've got at the top of my head for 2.2

- For errors, maybe some kind of scoped IAuthorizationErrorService, that the authZ services will report errors to with appropriate context.  This should enable apps to write an appropriate detailed authorization error.

- `UseAuthorize("policyName")` - middleware equivalent of `[Authorize("policyName")]`, would behave similarly, and also allow sideffect of setting httpContext.User if AuthenticationScheme is specified in the policy. TBD on what to do for failure, redirect to some access denied page which is able to display the appropriate context seems ideal.

**Comment by davidfowl** [comment]
I think we should close this issue and revisit this in 2.2 planning so we can scope it down.

**Comment by HaoK** [comment]
Close or leave it open for planning?

**Comment by hmvs** [comment]
Have you considered to add some ability to access to the list of policies which is run and which is failed. Also I thought it could be a helpful to have overload of method `context.Fail();` with some object which can be accessed later.

**Comment by natelaff** [comment]
Is this what currently prevents something such as the ability to show a specific page when a specific policy fails. For example, when a policy fails, redirect to a page that shows how to get that particular policy? (i.e. paying for a new membership level).

**Comment by tohsa** [comment]
Specifying a resource still wouldn't allow you to do something like pass a list of permissions to the attribute and there are too many possible combinations of permissions to make policies for every case. Maybe its not in the spirit of policies, but I don't understand how it is different than allowing a consumer to pass a list of roles.

All I want to be able to do is pass a list of permissions, call a microservice, and return a result, but it feels like I am going to have to write my own AuthorizeFilter at this point.
